testing/scenario_app/android/app/src/androidTest/java/dev/flutter/scenariosui/ScreenshotUtil.java

```diff
@@ -0,0 +1,257 @@
+   */
+  public static void capture(TestableFlutterActivity activity)
+      throws InterruptedException, ExecutionException, IOException {
+    // Yield and wait for the engine to render the first Flutter frame.
+    activity.waitUntilFlutterRendered();
+
+    // This method is called from the runner thread,
+    // so block the UI thread while taking the screenshot.
+    // UiThreadLocker locker = new UiThreadLocker();
+    // locker.lock();
+
+    // Screenshot.capture(view or activity) does not capture the Flutter UI.
+    // Unfortunately, it doesn't work with Android's `Surface` or `TextureSurface`.
+    //
+    // As a result, capture a screenshot of the entire device and then clip
+    // the status and action bars.
+    //
+    // Under the hood, this call is similar to `adb screencap`, which is used
+    // to capture screenshots.
+    final String testClass = TestNameDetector.getTestClass();
+    final String testName = TestNameDetector.getTestName();
+
+    runCallableOnUiThread(
+        new Callable<Void>() {
+          @Override
+          public Void call() {
+            Bitmap bitmap = Screenshot.capture().getBitmap();
+            // Remove the status and action bars from the screenshot capture.
+            bitmap =
+                Bitmap.createBitmap(
+                    bitmap,
+                    0,
+                    getStatusBarHeight(),
+                    bitmap.getWidth(),
+                    bitmap.getHeight() - getStatusBarHeight() - getActionBarHeight(activity));
+
+            final String screenshotName = String.format("%s__%s", testClass, testName);
+            // Write bitmap to the album.
+            try {
+              ScreenshotUtil.getInstance().writeBitmap(bitmap, screenshotName, testClass, testName);
+            } catch (IOException e) {
+              throw new RuntimeException(e);
+            }
+            return null;
+          }
+        });
+  }
+
+  /**
+   * Initializes the {@code com.facebook.testing.screenshot.internal.Album}.
+   *
+   * <p>Call this method from {@code AndroidJUnitRunner#onCreate}.
+   */
+  public static void onCreate(AndroidJUnitRunner runner, Bundle arguments) {
+    ScreenshotRunner.onCreate(runner, arguments);
+    ScreenshotUtil.getInstance().init();
+  }
+
+  /**
+   * Flushes the {@code com.facebook.testing.screenshot.internal.Album}.
+   *
+   * <p>Call this method from {@code AndroidJUnitRunner#onDestroy}.
+   */
+  public static void onDestroy() {
+    ScreenshotRunner.onDestroy();
+    ScreenshotUtil.getInstance().flush();
+  }
+
+  private static void runCallableOnUiThread(final Callable<Void> callable) {
+    if (Looper.getMainLooper().getThread() == Thread.currentThread()) {
+      try {
+        callable.call();
+      } catch (Exception e) {
+        e.printStackTrace();
+      }
+      return;
+    }
+    Handler handler = new Handler(Looper.getMainLooper());
+    final Object lock = new Object();
+    synchronized (lock) {
```

Comment:
are you really calling this method multiple times concurrently? Just synchronized (this)?

Reply:
This is static method, so it's a utility method.